File: phone/acg-ios/ACG/Screens/HomeScreen.swift
import SwiftUI

struct HomeScreen: View {
    @ObservedObject var world: World
    @State private var health: ApiClient.Health?
    @State private var ent: ApiClient.EnterpriseSnapshot?
    @State private var error: String?
    @State private var loading = false

    var body: some View {
        ScrollView {
            VStack(alignment: .leading, spacing: 12) {
                Text("⚒ AI Craftspeople Guild").font(.title2.bold())
                Text("ISA-95 live control-plane").foregroundStyle(.secondary)

                if let e = error { Text(e).foregroundStyle(.red) }

                if let h = health {
                    StatCard(title: "catalog · /api/health.json", rows: [
                        ("papers", "\(h.paperCount ?? 0)"),
                        ("members", "\(h.memberCount ?? 0)"),
                        ("api",     h.apiVersion ?? "—"),
                        ("updated", String((h.lastUpdated ?? "—").prefix(10))),
                    ])
                }
                if let e = ent {
                    StatCard(title: "enterprise · /runtime/tags.json", rows: [
                        ("papers", "\(e.papers)"), ("members", "\(e.members)"),
                        ("programs", "\(e.programs)"), ("runs", "\(e.runs)"),
                        ("tag edges", "\(e.tagEdges)"), ("authored lnks", "\(e.authoredLinks)"),
                    ])
                }
                StatCard(title: "local cache", rows: [
                    ("tags",      "\(world.tagDb.count())"),
                    ("last sync", world.tagDb.getMeta("last_full_sync").flatMap { formatAgo($0) } ?? "never"),
                ])

                HStack(spacing: 8) {
                    Button("Refresh") { Task { await refresh() } }.buttonStyle(.borderedProminent).disabled(loading)
                    Button("Sync tags") { Task { await syncTags() } }.buttonStyle(.bordered).disabled(loading)
                    if loading { ProgressView() }
                }
            }.padding()
        }.task { await refresh() }
    }

    private func refresh() async {
        loading = true; error = nil
        do { async let h = world.api.health(); async let e = world.api.enterprise()
            self.health = try await h; self.ent = try await e
        } catch { self.error = "\(error)" }
        loading = false
    }

    private func syncTags() async {
        loading = true; error = nil
        let r = await world.sync.pullAllTags()
        error = "✓ pulled \(r.pulled) tags · \(r.errors) errors · \(r.durationMs)ms"
        loading = false
    }

    private func formatAgo(_ msStr: String) -> String? {
        guard let ms = Double(msStr) else { return nil }
        let secs = Int((Date().timeIntervalSince1970 * 1000 - ms) / 1000)
        switch secs {
        case ..<60:    return "\(secs)s ago"
        case ..<3600:  return "\(secs / 60)m ago"
        case ..<86400: return "\(secs / 3600)h ago"
        default:       return "\(secs / 86400)d ago"
        }
    }
}

struct StatCard: View {
    let title: String
    let rows: [(String, String)]
    var body: some View {
        VStack(alignment: .leading, spacing: 6) {
            Text(title.uppercased())
                .font(.system(size: 10, weight: .bold, design: .monospaced))
                .kerning(2).foregroundStyle(.green)
            ForEach(rows, id: \.0) { k, v in
                HStack { Text(k).foregroundStyle(.secondary); Spacer(); Text(v).bold() }
            }
        }
        .padding(14)
        .frame(maxWidth: .infinity, alignment: .leading)
        .background(Color(white: 0.1))
        .clipShape(RoundedRectangle(cornerRadius: 8))
    }
}

File: phone/acg-android/app/src/main/java/com/aicraftspeopleguild/acg/ui/HomeScreen.kt
package com.aicraftspeopleguild.acg.ui

import androidx.compose.runtime.*
import androidx.compose.ui.Modifier
import com.aicraftspeopleguild.acg.data.ApiClient
import com.aicraftspeopleguild.acg.data.Health
import com.aicraftspeopleguild.acg.data.SyncEngine
import com.aicraftspeopleguild.acg.data.TagDb
import com.aicraftspeopleguild.acg.udt.*
import com.aicraftspeopleguild.acg.ui.organism.StatScreen
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

/**
 * Whole screen is declared as a StatScreenParams tree — no bespoke
 * layout code. Change a value, add a card, or flip a button variant
 * by editing data, not Compose.
 */
@Composable
fun HomeScreen(api: ApiClient, tagDb: TagDb, sync: SyncEngine, modifier: Modifier = Modifier) {
    var health by remember { mutableStateOf<Health?>(null) }
    var ent by remember { mutableStateOf<ApiClient.EnterpriseSnapshot?>(null) }
    var status by remember { mutableStateOf<StatusLineParams?>(null) }
    var loading by remember { mutableStateOf(false) }
    var localCount by remember { mutableStateOf(tagDb.countTags()) }
    var lastSync by remember { mutableStateOf(tagDb.getMeta("last_full_sync")) }
    val scope = rememberCoroutineScope()

    fun refresh() = scope.launch {
        loading = true; status = null
        try {
            withContext(Dispatchers.IO) {
                health = api.health(); ent = api.enterprise()
            }
        } catch (e: Exception) {
            status = StatusLineParams("✗ ${e.message}", Severity.ERR)
        }
        loading = false
    }

    fun syncTags() = scope.launch {
        loading = true; status = null
        try {
            val r = sync.pullAllTags()
            localCount = tagDb.countTags()
            lastSync = tagDb.getMeta("last_full_sync")
            status = StatusLineParams("✓ pulled ${r.tagsPulled} tags · ${r.errors} errors · ${r.durationMs}ms", Severity.OK)
        } catch (e: Exception) {
            status = StatusLineParams("✗ sync: ${e.message}", Severity.ERR)
        }
        loading = false
    }

    LaunchedEffect(Unit) { refresh() }

    StatScreen(
        modifier = modifier,
        p = StatScreenParams(
            title    = "⚒ AI Craftspeople Guild",
            subtitle = "ISA-95 live control-plane",
            status   = status,
            cards    = listOfNotNull(
                health?.let {
                    StatCardParams("catalog · /api/health.json", listOf(
                        StatRowParams("papers",  it.paperCount.toString()),
                        StatRowParams("members", it.memberCount.toString()),
                        StatRowParams("api",     it.apiVersion ?: "—"),
                        StatRowParams("updated", it.lastUpdated?.take(10) ?: "—"),
                    ))
                },
                ent?.let {
                    StatCardParams("enterprise · /runtime/tags.json", listOf(
                        StatRowParams("papers",        it.papers.toString()),
                        StatRowParams("members",       it.members.toString()),
                        StatRowParams("programs",      it.programs.toString()),
                        StatRowParams("runs",          it.runs.toString()),
                        StatRowParams("tag edges",     it.tagEdges.toString()),
                        StatRowParams("authored lnks", it.authoredLinks.toString()),
                    ))
                },
                StatCardParams("local cache", listOf(
                    StatRowParams("tags",      localCount.toString()),
                    StatRowParams("last sync", lastSync?.toLongOrNull()?.let { formatAgo(it) } ?: "never"),
                )),
            ),
            actionRow = ActionRowParams(
                busy = loading,
                buttons = listOf(
                    ActionButtonParams("Refresh",   ButtonVariant.PRIMARY,   enabled = !loading),
                    ActionButtonParams("Sync tags", ButtonVariant.SECONDARY, enabled = !loading),
                ),
            ),
        ),
        onAction = { label ->
            when (label) {
                "Refresh"   -> refresh()
                "Sync tags" -> syncTags()
            }
        },
    )
}

private fun formatAgo(epochMs: Long): String {
    val secs = (System.currentTimeMillis() - epochMs) / 1000
    return when {
        secs < 60    -> "${secs}s ago"
        secs < 3600  -> "${secs / 60}m ago"
        secs < 86400 -> "${secs / 3600}h ago"
        else         -> "${secs / 86400}d ago"
    }
}
